corehq/apps/locations/models.py

```diff
@@ -27,10 +28,36 @@ class SQLLocation(MPTTModel):
     latitude = models.DecimalField(max_digits=20, decimal_places=10, null=True)
     longitude = models.DecimalField(max_digits=20, decimal_places=10, null=True)
     parent = TreeForeignKey('self', null=True, blank=True, related_name='children')
-    products = models.ManyToManyField(SQLProduct, null=True)
+
+    # Use getter and setter below to access this value
+    # since stocks_all_products can cause an empty list to
+    # be what is stored for a location that actually has
+    # all products available.
+    _products = models.ManyToManyField(SQLProduct, null=True)
+    stocks_all_products = models.BooleanField(default=True)
 
     supply_point_id = models.CharField(max_length=255, db_index=True, unique=True, null=True)
 
+    @property
+    def products(self):
+        """
+        If there are no products specified for this location, assume all
+        products for the domain are relevant.
+        """
+        if self.stocks_all_products:
+            return SQLProduct.by_domain(self.domain)
+        else:
+            return self._products.all()
+
+    @products.setter
+    def products(self, value):
+        # this will set stocks_all_products to true if the user
+        # has added all products in the domain to this location
+        self.stocks_all_products = (set(value) ==
+                                    set(SQLProduct.by_domain(self.domain)))
```

Comment:
Basically, if the user selects all products for their location, any new products on that domain will also show up.  We debated this a bit, and this seems like the right combination of ease-of-implementation and intuitiveness for the user.